lib/user_space/mempressure.c

```diff
@@ -0,0 +1,520 @@
+ * 	component objects[ beck_tool ] when found atomic  couter
+ * 		 has been set, it executes its bottom half.
+ * 	component[ beck tool ]  event_bh  call builder_fini/ini.
+ *
+ * m0_mempressure_mod_fini() 
+ * 	Unloads mempressure module.
+ *
+ * m0_mempressure_cb_add()
+ * 	For subscribing for event callback event_th[ top half ]
+ *
+ * m0_mempressure_cb_del()
+ * 	Unsubscribe event_th from mempresure.
+ *
+ * m0_mempressure_get()
+ * 	Get current level[ mem pressure ]
+ *
+ * @todo Interface
+ *  current implementation done for memory subsystem.
+ *  Need to generalize for other subsystem resources.
+ *
+ *
+ * @{
+ */
+
+static int  thread_start(struct mempressure_obj *lobj);
+static void listener_thread(struct mempressure_obj *lobj);
+static void close_fd_on_exit(struct mempressure_obj *lobj);
+
+static void mp_event_th_run(struct m0_mempressure *loc);
+static void event_group_lock(struct m0_mempressure *mp_obj);
+static void event_group_unlock(struct m0_mempressure *loc);
+static int  fd_to_level(struct mempressure_obj *mobj, int fd);
+
+static int path_snprintf(char *buf, size_t buflen,
+				const char *format, ...);
+static int system_do(char *buf, size_t buflen, const char *format, ...);
+
+M0_TL_DESCR_DEFINE(runmq, "runq mp", static,
+			struct m0_mempressure_cb, mc_linkage,
+			mp_magic, M0_MP_MAGIC, M0_MP_RUNQ_MAGIC);
+M0_TL_DEFINE(runmq, static, struct m0_mempressure_cb);
+
+enum {
+	MAX_EVENTS = 3,
+	SLEN       = 32,
+	TIME_OUT   = 20000,
+};
+
+/**
+ * cgroup files paths and groupname.
+ */
+static const char  mempath[]           = "/sys/fs/cgroup/memory/";
+static const char  m_pressure_level[]  = "memory.pressure_level";
+static const char  cg_event_controll[] = "cgroup.event_control";
+static const char *level_name[] = {
+	"low",
+	"medium",
+	"critical",
+};
+static char        c_group[SLEN];
+
+/**
+ * variadic path generating function.
+ */
+static int path_snprintf(char *buf, size_t buflen, const char *format, ...)
+{
+	va_list ap;
+	int     len;
+
+	va_start(ap, format);
+	len = vsnprintf(buf, buflen, format, ap);
+	va_end(ap);
+	if (len <= 0 || len >= buflen)
+		return -1;
+	return len;
+}
+
+/**
+ * variadic command execute function.
+ */
+static int system_do(char *buf, size_t buflen, const char *format, ...)
+{
+	va_list ap;
+	int     res;
+
+	va_start(ap, format);
+	res = vsnprintf(buf, buflen, format, ap);
+	va_end(ap);
+	if (res == -1) {
+		M0_LOG(M0_ERROR, "buffer error");
+		return -1;
+	}
+	res = system(buf);
+	if (res != 0)
+		return -1;
+	return res;
+}
+
+/**
+ * group lock to protect subscriber event queue.
+ */
+static void event_group_lock(struct m0_mempressure *mp_obj)
+{
+	m0_mutex_lock(&mp_obj->mp_gr_lock);
+}
+
+/**
+ * Publishes a new event to all the subscribers.
+ */
+static void mp_event_th_run(struct m0_mempressure *mp)
+{
+	struct m0_mempressure_cb *icb;
+
+	M0_ENTRY();
+	M0_LOG(M0_DEBUG, "mempressure event th for level %d.",
+	       mp->mp_cur_level);
+	m0_tl_for(runmq, &mp->mp_runmq, icb) {
```

Comment:
Locked needs to be held while iterating the list.
If it is already held, please assert() it here.